File: Den.Dev.Grunt/Den.Dev.Grunt/Core/Modules/TextModerationModule.cs
// <copyright file="TextModerationModule.cs" company="Den Delimarsky">
// Developed by Den Delimarsky.
// Den Delimarsky licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
// The underlying API powering Den.Dev.Grunt is managed by Halo Studios and Microsoft. This wrapper is not endorsed by Halo Studios or Microsoft.
// </copyright>

using System.Threading.Tasks;
using Den.Dev.Grunt.Core.Foundation;
using Den.Dev.Grunt.Endpoints;
using Den.Dev.Grunt.Models;
using Den.Dev.Grunt.Models.HaloInfinite;

namespace Den.Dev.Grunt.Core.Modules
{
    /// <summary>
    /// Module for text moderation related API operations.
    /// </summary>
    public class TextModerationModule : ModuleBase
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="TextModerationModule"/> class.
        /// </summary>
        /// <param name="client">The client instance to use for API requests.</param>
        internal TextModerationModule(ClientBase client)
            : base(client, HaloCoreEndpoints.TextOrigin)
        {
        }

        /// <summary>
        /// Gets a specific moderation proof signing key.
        /// </summary>
        /// <param name="keyId">Key ID. Full list can be obtained by a call to GetSigningKeys.</param>
        /// <returns>An instance of Key containing a single signing key data if request was successful. Return value is null otherwise.</returns>
        public async Task<HaloApiResultContainer<Key, RawResponseContainer>> GetSigningKey(string keyId)
        {
            return await this.GetAsync<Key>(
                $"/hi/moderation-proof-keys/{keyId}",
                useSpartanToken: false);
        }

        /// <summary>
        /// Gets a list of available moderation proof signing keys.
        /// </summary>
        /// <returns>An instance of ModerationProofKeys containing signing key data if request was successful. Return value is null otherwise.</returns>
        public async Task<HaloApiResultContainer<ModerationProofKeys, RawResponseContainer>> GetSigningKeys()
        {
            return await this.GetAsync<ModerationProofKeys>(
                "/hi/moderation-proof-keys",
                useSpartanToken: false);
        }
    }
}

File: src/node/src/modules/halo-infinite/text-moderation.module.ts
import { ModuleBase } from '../base/module-base';
import type { ClientBase } from '../../clients/base/client-base';
import type { HaloApiResult } from '../../models/common/api-result';
import { HALO_CORE_ENDPOINTS } from '../../endpoints/halo-core-endpoints';

/**
 * Signing key for moderation proofs.
 */
export interface SigningKey {
  /** Key type */
  kty?: string;
  /** Key use */
  use?: string;
  /** Key ID */
  kid?: string;
  /** Key modulus (RSA) */
  n?: string;
  /** Key exponent (RSA) */
  e?: string;
  /** X.509 certificate chain */
  x5c?: string[];
}

/**
 * Container for moderation proof signing keys.
 */
export interface ModerationProofKeys {
  /** List of signing keys */
  keys?: SigningKey[];
}

/**
 * Text Moderation module for moderation-related APIs.
 *
 * @example
 * ```typescript
 * // Get all signing keys
 * const keys = await client.textModeration.getSigningKeys();
 *
 * // Get a specific signing key
 * const key = await client.textModeration.getSigningKey('key-id');
 * ```
 */
export class TextModerationModule extends ModuleBase {
  constructor(client: ClientBase) {
    super(client, HALO_CORE_ENDPOINTS.TEXT_ORIGIN);
  }

  /**
   * Get a specific moderation proof signing key.
   *
   * @param keyId - Key ID (can be obtained from getSigningKeys)
   * @returns Signing key data
   */
  getSigningKey(keyId: string): Promise<HaloApiResult<SigningKey>> {
    this.assertNotEmpty(keyId, 'keyId');
    return this.get<SigningKey>(`/hi/moderation-proof-keys/${keyId}`, {
      useSpartanToken: false,
    });
  }

  /**
   * Get all available moderation proof signing keys.
   *
   * @returns Container with all signing keys
   */
  getSigningKeys(): Promise<HaloApiResult<ModerationProofKeys>> {
    return this.get<ModerationProofKeys>('/hi/moderation-proof-keys', {
      useSpartanToken: false,
    });
  }
}
